### PR TITLE
state: replace workers package with worker.Runner

### PR DESCRIPTION
**Pull request by rogpeppe**
This fixes https://bugs.launchpad.net/juju/+bug/1669540

Instead of the custom state/workers package with, we use the existing
worker.Runner implementation. This raises the issue of what to do
when a worker cannot be acquired. The obvious solution to that is
to change the worker methods to return an error, but that turns out
to be a huge change because it means that every single Watch method
needs to return an error, and hence all the callers of those.
Instead of causing massive disruption to the code base like that, we
instead observe that watchers already have a way of returning an error -
via the Wait method. So we implement a few extra functions that can
return already-dead workers in the relevant packages (state/watcher,
state/presence and worker/lease) and use those instead.

### CODE REVIEW

File: worker/workertest/check.go [hunk 1]
```diff
@@ -49,6 +49,8 @@ func CheckAlive(c *gc.C, w worker.Worker) {
 func CheckKilled(c *gc.C, w worker.Worker) error {
 	wait := make(chan error, 1)
 	go func() {
+		c.Logf("worker %T", w)
```
Comment (wallyworld) [inline worker/workertest/check.go]: Can we remove these, they seem just for debugging

File: dependencies.tsv [hunk 2]
```diff
@@ -1,56 +1,31 @@
-github.com/Azure/azure-sdk-for-go	git	902d95d9f311ae585ee98cfd18f418b467d60d5a	2016-07-20T05:16:58Z
```
Comment (wallyworld) [inline dependencies.tsv]: Woah. These deletions seem wrong.

File: worker/workertest/check.go (hunk 1 above)
Comment (rogpeppe) [inline worker/workertest/check.go]: done

File: state/resources_mongo.go [hunk 3]
```diff
@@ -213,7 +213,7 @@ func newResolvePendingResourceOps(pending storedResource, exists bool) []txn.Op
 		Resource:      newRes.Resource.Resource,
 		id:            newRes.ID,
 		applicationID: newRes.ApplicationID,
-		lastPolled:    time.Now().UTC(),
+		lastPolled:    time.Now().AddDate(0, 0, 0).UTC(),
```
Comment (wallyworld) [inline state/resources_mongo.go]: please add comment here and elsewhere to explain the AddData() usage
maybe even a todo to fix it properly

File: state/unit_test.go [hunk 4]
```diff
@@ -1761,17 +1761,19 @@ snapshot:
...
 			actionName:   "juju-run",
 			givenPayload: map[string]interface{}{"command": "allyourbasearebelongtous"},
-			errString:    `validation failed: (root) : "timeout" property is missing and required, given {"command":"allyourbasearebelongtous"}`,
+			errString:    `validation failed: \(root\) : "timeout" property is missing and required, given \{"command":"allyourbasearebelongtous"\}`,
 		},
 		{
 			actionName:   "juju-run",
 			givenPayload: map[string]interface{}{"timeout": 5 * time.Second},
-			errString:    `validation failed: (root) : "command" property is missing and required, given {"timeout":5e+09}`,
+			// Note: in Go 1.8 the representation of large numbers in JSON changed
```
Comment (wallyworld) [inline state/unit_test.go]: Yes! I hit this issue but hadn't fixed it yet. Thanks!

File: worker/lease/manager.go [hunk 5]
```diff
@@ -21,6 +21,19 @@ var logger = loggo.GetLogger("juju.worker.lease")
...
 var errStopped = errors.New("lease manager stopped")
 
+// NewDeadManager returns a manager that's already dead
+// and always returns the given error.
+func NewDeadManager(err error) *Manager {
+	var m Manager
+	catacomb.Invoke(catacomb.Plan{
+		Site: &m.catacomb,
+		Work: func() error {
+			return errors.New("hello")
```
Comment (wallyworld) [inline worker/lease/manager.go]: this doesn't seem to match the doc comment
shouldn't it be "return err"?

File: state/multiwatcher.go [hunk 6]
```diff
@@ -183,6 +184,15 @@ func newStoreManagerNoRun(backing Backing) *storeManager {
 	}
 }
 
+// newDeadStoreManager returns a store manager instance
+// that is already dead and always returns the given error.
+func newDeadStoreManager(err error) *storeManager {
+	var m storeManager
+	m.tomb.Kill(errors.Trace(err))
```
Comment (wallyworld) [inline state/multiwatcher.go]: ISTM that the callers are passing in errors.Trace(err)
So we need to just use Kill(err) here?

File: state/presence/presence.go [hunk 7]
```diff
@@ -130,6 +130,15 @@ type Change struct {
 	Alive bool
 }
 
+// NewDeadWatcher returns a new watcher that is already dead
+// and always returns the given error from its Err method.
+func NewDeadWatcher(err error) *Watcher {
+	var w Watcher
+	w.tomb.Kill(errors.Trace(err))
```
Comment (wallyworld) [inline state/presence/presence.go]: ditto about errors.Trace(err)

File: worker/lease/manager.go (hunk 5 above)
Comment (rogpeppe) [inline worker/lease/manager.go]: Indeed, fixed.

File: state/multiwatcher.go (hunk 6 above)
Comment (rogpeppe) [inline state/multiwatcher.go]: I think it's good to trace the error whereever it's passed on.

File: state/presence/presence.go (hunk 7 above)
Comment (rogpeppe) [inline state/presence/presence.go]: ditto about tracing everywhere.

File: state/resources_mongo.go (hunk 3 above)
Comment (rogpeppe) [inline state/resources_mongo.go]: added comment.

**Comment by rogpeppe** [comment]
$$merge$$

**Comment by rogpeppe** [comment]
$$merge$$

**Comment by rogpeppe** [comment]
$$merge$$

**Pull request merged**